### PR TITLE
Added more robust shutdown code for our dmJobThread implementation

### PR DESCRIPTION
**Pull request by JCash**
This fixes an issue where it was possible for get a dead lock when shutting down the job thread system.

## PR checklist

* [ ] Code
	* [ ] Add engine and/or editor unit tests.
	* [ ] New and changed code follows the overall code style of existing code
	* [ ] Add comments where needed
* [ ] Documentation
	* [ ] Make sure that API documentation is updated in code comments
	* [ ] Make sure that manuals are updated (in github.com/defold/doc)
* [ ] Prepare pull request and affected issue for automatic release notes generator
	* [ ] Pull request - Write a message that explains what this pull request does. What was the problem? How was it solved? What are the changes to APIs or the new APIs introduced? This message will be used in the generated release notes. Make sure it is well written and understandable for a user of Defold.
	* [ ] Pull request - Write a pull request title that in a sentence summarises what the pull request does. Do not include "Issue-1234 ..." in the title. This text will be used in the generated release notes.
	* [ ] Pull request - Link the pull request to the issue(s) it is closing. Use on of the [approved closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).
	* [ ] Affected issue - Assign the issue to a project. Do not assign the pull request to a project if there is an issue which the pull request closes.
	* [ ] Affected issue - Assign the "breaking change" label to the issue if introducing a breaking change.
	* [ ] Affected issue - Assign the "skip release notes" is the issue should not be included in the generated release notes.

----------

Example of a well written PR description:

1. Start with the user facing changes. This will end up in the release notes.
1. Add one of the GitHub approved closing keywords
1. Optionally also add the technical changes made. This is information that might help the reviewer. It will not show up in the release notes. Technical changes are identified by a line starting with one of these:
   1. `### Technical changes` 
   1. `Technical changes:`
   2. `Technical notes:`

```
There was a anomaly in the carbon chroniton propeller, introduced in version 8.10.2. This fix will make sure to reset the phaser collector on application startup.

Fixes #1234

### Technical changes
* Pay special attention to line 23 of phaser_collector.clj as it contains some interesting optimizations
* The propeller code was not taking into account a negative phase.
```

### CODE REVIEW

File: engine/dlib/src/dlib/job_thread.cpp [hunk 1]
```diff
@@ -53,7 +53,7 @@ struct JobThreadContext
 #if defined(DM_HAS_THREADS)
     dmMutex::HMutex                         m_Mutex;
     dmConditionVariable::HConditionVariable m_WakeupCond;
-    int32_atomic_t                          m_Run;
+    bool                                    m_Run;
```
Comment (JCash) [inline engine/dlib/src/dlib/job_thread.cpp:56]: No need for atomics if we already use a mutex.

File: engine/dlib/src/dlib/job_thread.cpp [hunk 2]
```diff
@@ -89,16 +89,19 @@ static void PutDone(JobThreadContext* ctx, JobItem* item)
 static void JobThread(void* _ctx)
 {
     JobThreadContext* ctx = (JobThreadContext*)_ctx;
-    while (dmAtomicGet32(&ctx->m_Run) != 0)
+    while (true)
     {
-        JobItem item;
+        JobItem item = {};
```
Comment (JCash) [inline engine/dlib/src/dlib/job_thread.cpp:94]: Between the loop check and the dmConditionVariable::Wait(), the main thread set `run=0` and also signalled the wakeup condition...

File: engine/dlib/src/dlib/job_thread.cpp [hunk 3]
```diff
@@ -89,16 +89,19 @@ static void PutDone(JobThreadContext* ctx, JobItem* item)
...
         {
             DM_MUTEX_SCOPED_LOCK(ctx->m_Mutex);
+
+            if (!ctx->m_Run)
+                break;
+
             while(ctx->m_Work.Empty())
             {
                 dmConditionVariable::Wait(ctx->m_WakeupCond, ctx->m_Mutex);
```
Comment (JCash) [inline engine/dlib/src/dlib/job_thread.cpp:103]: Once we got here, the signal had already been sent which meant that the wait would then stall forever, as there wouldn't be any new signal coming.

File: engine/dlib/src/dlib/job_thread.cpp [hunk 4]
```diff
@@ -152,14 +155,12 @@ void Destroy(HContext context)
...
         DM_MUTEX_SCOPED_LOCK(context->m_ThreadContext.m_Mutex);
-        // Wake up the worker threads so it can exit and allow us to join
-        for (int i = 0; i < context->m_Threads.Size(); ++i)
-        {
-            dmConditionVariable::Signal(context->m_ThreadContext.m_WakeupCond);
-        }
+
+        context->m_ThreadContext.m_Run = false;;
+
+        dmConditionVariable::Broadcast(context->m_ThreadContext.m_WakeupCond);
```
Comment (JCash) [inline engine/dlib/src/dlib/job_thread.cpp:163]: Better to wakeup all threads at once.

File: engine/dlib/src/dlib/job_thread.cpp [hunk 5]
```diff
@@ -152,14 +155,12 @@ void Destroy(HContext context)
...
-    dmAtomicStore32(&context->m_ThreadContext.m_Run, 0);
     {
         DM_MUTEX_SCOPED_LOCK(context->m_ThreadContext.m_Mutex);
-        // Wake up the worker threads so it can exit and allow us to join
-        for (int i = 0; i < context->m_Threads.Size(); ++i)
-        {
-            dmConditionVariable::Signal(context->m_ThreadContext.m_WakeupCond);
-        }
+
+        context->m_ThreadContext.m_Run = false;;
```
Comment (JCash) [inline engine/dlib/src/dlib/job_thread.cpp:161]: Now we know that the Run and Wait() are within the same protected scope on the thread and here as well.

**Review by Jhonnyg — APPROVED** [review]
Lgtm but there’s no tests, do we want that perhaps?

**Comment by JCash** [comment]
I'd say yes, but I'm not sure how to write such a test that actually catches/tests the issue in a 100% manner?

**Pull request merged**